contracts/Payments.sol

```diff
@@ -0,0 +1,96 @@
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@api3/airnode-protocol-v1/contracts/dapis/DapiReader.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/utils/Counters.sol";
+
+contract Payments is ERC721, DapiReader, Ownable {
+    using Counters for Counters.Counter;
+
+    Counters.Counter private _tokenIdCounter;
+
+    ERC20 public _ERC20;
+    mapping(address => bytes32) public tokenDapiMapping;
+    mapping(uint256 => uint256) public TokenIDtoPrice; 
+    address[] allowedTokens;
+
+constructor(address _dapiServer) DapiReader(_dapiServer) ERC721("Payment Receipt", "PRT") {
+
+    }
+    // @dev set dAPI names for tokens
+    function setDapiName(address token, bytes32 DapiName)
+        public
+        onlyOwner
+    {
+        tokenDapiMapping[token] = DapiName;
+    }
+    // @dev add allowed tokens
+    function addAllowedTokens(address token) public onlyOwner {
+        allowedTokens.push(token);
+    }
+    // @dev check if token is allowed
+    function tokenIsAllowed(address token) public view returns (bool) {
+        for (uint256 i = 0; i < allowedTokens.length; i++) {
+            if (allowedTokens[i] == token) {
+                return true;
+            }
+        }
+        return false;
+    }
+    // @dev get token price from dAPI name
+    function getTokenPrice(address token) public view returns (uint256 tokenPriceUint256) {
+        bytes32 DapiName = tokenDapiMapping[token];
+        int224 value = IDapiServer(dapiServer).readDataFeedValueWithDapiName(
+                DapiName
+        );
+        uint224 tokenPriceUint224 = uint224(value);
+        tokenPriceUint256 = tokenPriceUint224;
+        return tokenPriceUint256;
+    }
+
+    // @dev make the receipt
+    function makeReceipt(uint256 tokenId, uint256 price) internal returns (uint256) {
+        TokenIDtoPrice[tokenId] = price;
+        return (price);
+    }
+
+    // @dev make the payment
+    function Payment(address token, uint256 _tokenAmount) public returns(uint256) {
+        require(_tokenAmount > 0, "amount cannot be 0");
+        require(tokenIsAllowed(token), "token not allowed");
+        if (tokenIsAllowed(token)) {
+            uint256 tokenId = _tokenIdCounter.current();
+            _tokenIdCounter.increment();
+            uint256 decimals = ERC20(token).decimals();
+            uint256 _usdValue = (getTokenPrice(token) * _tokenAmount)/10**decimals;
+            ERC20(token).transferFrom(msg.sender, address(this), _tokenAmount);
+            _safeMint(msg.sender, tokenId);
+            uint256 receipt = makeReceipt(tokenId, _usdValue);
+            return receipt;
+    }
+    }
+
+    // @dev check receipt withn tokenId
+    function checkReceipt(uint256 tokenId) public view returns (uint256) {
+        return (TokenIDtoPrice[tokenId]);
+    }
+    
+    // @dev get all payments in USD
+    function getTotalPayments() public view returns(uint256) {
```

Comment:
1. This should be parsing the tokenIds which already has the USD value at the time of the payment but that might still be a bad idea
2. This gets the price when you call the function not at the time of payment
3. There isn't a clean way to do this via a mapping, maybe call it `getContractBalance(address token)` and get the USD balance of the contract for a specific token

Reply:
- Yeah but using a mapping would make `getTotalPayments()` and `ownerWithdrawFunds()` unusable for users if they have to add the address of each token manually to check/withdraw.

- Can just rename `getTotalPayments()` to `getContractBalance()`?